skip Positron Proxy for fastapi and streamlit apps on PWB

sharon-wang opened the pull request:
- addresses: https://github.com/posit-dev/positron/issues/5027 and #5026 
- skip the Positron Proxy when running fastapi and streamlit apps in Positron on Workbench
- add log output channel for Positron Proxy for easier debugging
- add typescript dev dependency to Positron Run App package.json to make `vsce package` available for that extension

### QA Notes

I've been testing with the apps from https://github.com/posit-dev/qa-example-content/tree/main/workspaces/python_apps.

- Positron on Workbench (Linux Ubuntu x86_64): Dash, Fastapi, Flask, Gradio, Streamlit all good ✅ 
    - Note that I sometimes have to refresh the Viewer for the Fastapi app before it works, as the app may be opened in the Viewer before the app is actually ready
- Positron Server Web (Mac): Dash, Fastapi, Flask, Gradio, Streamlit all good ✅ 
- Positron Desktop (Mac): Dash, Fastapi, Flask, Gradio, Streamlit all good ✅

sharon-wang committed:
add log output channel for positron proxy

sharon-wang committed:
fix handling of custom urlPath in positron-run-app

sharon-wang committed:
remove trailing slashes and add trace logging

sharon-wang committed:
add workbench-specific handling to determine the proxy targetOrigin

sharon-wang committed:
remove removal of trailing slash from target origin

I thought it was fixing something, but it turns out it's not, and actually it breaks some frameworks such as Dash.

sharon-wang commented on extensions/positron-run-app/src/api.ts:
```diff
@@ -353,18 +353,27 @@ async function previewUrlInExecutionOutput(execution: vscode.TerminalShellExecut
 
 	// Preview the external URI.
-	positron.window.previewUrl(proxyInfo.externalUri);
+	positron.window.previewUrl(previewUri);
```
@seeM When running Fastapi in Positron on Workbench, sometimes the app url is previewed in the Viewer slightly too early, before the app server is ready. To get around this, a user can refresh the Viewer, but that is not ideal if it happens every time.

Do you have thoughts on how to get around this issue? Maybe we can ping the uri to check if it's responsive before displaying it in the Viewer?

sharon-wang replied:
Maybe https://github.com/posit-dev/positron/issues/4276 is describing what I'm seeing?

seeM replied:
I assume that uvicorn would only print the URL to stdout once the server is ready to receive connections (I haven't confirmed that). Is the delay happening at the Positron Proxy layer? If so, maybe there's something we could expose from the Positron Proxy extension i.e. an `await`able or an event that we could listen to here?

sharon-wang replied:
I looked into this a bit more and it does not seem to be in the Positron Proxy layer, as the issue still occurs even if we don't use the Positron Proxy.

For fastapi with uvicorn, we may want to wait for `Application startup complete` in the terminal output before we open the app at the url. We happen to be just a little quick to open the url/slow to startup the app on Workbench, such that the fastapi app is not quite ready yet.

Perhaps this would be a separate property in the terminal/debug config for some terminal text that indicates the app is ready, before we extract the app url and view it. I'll open a separate issue for this.

sharon-wang replied:
Opened https://github.com/posit-dev/positron/issues/5187

sharon-wang commented on extensions/positron-run-app/src/api.ts:
```diff
@@ -353,18 +353,27 @@ async function previewUrlInExecutionOutput(execution: vscode.TerminalShellExecut
 
-	log.debug(`Viewing app at local uri: ${localUri} with external uri ${proxyInfo.externalUri.toString()}`);
+	// Determine the target origin based on whether the server is running in PWB.
+	const runningInPWB = process.env.RS_SERVER_URL ? true : false;
```
Maybe we want this check in a command in case we end up using it in a bunch of places?

seeM replied:
That seems reasonable.

sharon-wang replied:
I'm going to defer this to a future task.

sharon-wang committed:
add websocket proxy req logging and explicit upgrade handling

sharon-wang committed:
override websocket proxy request origin header

seeM commented on extensions/positron-run-app/src/api.ts:
```diff
@@ -353,18 +353,27 @@ async function previewUrlInExecutionOutput(execution: vscode.TerminalShellExecut
-	log.debug(`Viewing app at local uri: ${localUri} with external uri ${proxyInfo.externalUri.toString()}`);
+	// Determine the target origin based on whether the server is running in PWB.
+	const runningInPWB = process.env.RS_SERVER_URL ? true : false;
+	const targetOrigin = runningInPWB ? localBaseUri : localUri;
```
Interesting. I'm curious why the target origin differs here for workbench

sharon-wang replied:
It's a bit of a hack as I couldn't get `UVICORN_ROOT_PATH` or `--root-path` to work for fastapi in Positron Server Web or Desktop. I haven't gotten to the bottom of why it's working on Workbench.

The result on Positron Server Web without this handling is that the `openapi.json` file is getting loaded from the root `/openapi.json` instead of from e.g. `proxy/1234/openapi.json`, so we get:

<img width="1208" alt="positron_web_fastapi_openapi_notfound" src="https://github.com/user-attachments/assets/93e9cbd2-96bc-441f-8388-f18fc564d347">

sharon-wang replied:
I'm gonna move this PR to Draft while I try a few more things to see if we can avoid this custom handling.

seeM reviewed (dismissed):
I haven't setup a workbench dev environment so I haven't tested the changes, but code changes look good. We should try to fix the fastapi preview if possible but don't think it has to block this PR.

sharon-wang committed:
WIP: extra handling in positron proxy for websockets

sharon-wang committed:
add  debug logging for websocket upgrades

sharon-wang committed:
try to rewrite the proxy with http-proxy

proxying kind of works, but getting:

Refused to apply style from 'path/to/style.css' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.

sharon-wang committed:
remove fastapi workbench-specific workaround

sharon-wang committed:
remove fastapi workbench-specific workaround

sharon-wang committed:
skip positron proxy for fastapi apps on PWB

sharon-wang committed:
add typescript dep to positron-run-app

this allows us to run `vsce package` in the `extensions/positron-run-app` directory to generate a VSIX which can be loaded into Positron for faster extension dev iteration when running a built reh-web version of Positron.

sharon-wang committed:
update function doc and match style with other functions

sharon-wang commented on extensions/positron-run-app/package.json:
```diff
@@ -39,7 +39,8 @@
     "@types/node": "^22.5.4",
     "@types/sinon": "^17.0.3",
-    "@types/sinon-test": "^2.4.6"
+    "@types/sinon-test": "^2.4.6",
+    "typescript": "^4.5.5"
```
this allows us to run `vsce package` in the `extensions/positron-run-app` directory to generate a VSIX, which can be loaded into Positron. This is helpful for faster extension dev iteration when running a pre-built reh-web version of Positron.

sharon-wang committed:
disable positron proxy for streamlit on PWB

sharon-wang committed:
Revert "try to rewrite the proxy with http-proxy"

This reverts commit 07b544b6e1ce4b44aeb5e3795a0cc8503a73c2e5.

sharon-wang committed:
Revert "add  debug logging for websocket upgrades"

This reverts commit 7e15c260efc3a1e5ac504b1a18179ee032afb527.

sharon-wang committed:
Revert "WIP: extra handling in positron proxy for websockets"

This reverts commit aeab1f2323bd3a85f1f9e491c40fca3793686d87.

sharon-wang committed:
Revert "override websocket proxy request origin header"

This reverts commit 2e9b4adefb64cfb15f5b652aea0a42f3ba271e3c.

sharon-wang committed:
Revert "add websocket proxy req logging and explicit upgrade handling"

This reverts commit e4fa27e2635dce05a87c47fe01f60348a3636a7d.
